### PR TITLE
feat: Add iterators over different types of columns in Schema

### PR DESCRIPTION
**Pull request by alamb**
This is related to #811 where I am moving `read_filter` into the gRPC planner, and is easily separable. I need to be able to iterate over tags / fields / timestamps easily so I can create selections in the plan in some straightforward way. 

This may also come in handy for some of the work that @NGA-TRAN  is doing on https://github.com/influxdata/influxdb_iox/pull/875

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: data_types/src/schema.rs [hunk 1]
```diff
@@ -567,30 +589,131 @@ impl From<&InfluxColumnType> for ArrowDataType {
     }
 }
 
+/// Trait for things that can filter for specific types of columns
+/// while iterating
+pub trait ColumnFilter: Default {
```
Comment (alamb) [inline data_types/src/schema.rs:594]: I would love any Rust expert commentary about these iterators and if there is a better / more Rust idiomatic way

**Comment by e-dard** [comment]
I think I'm probably missing the big picture here, but from what I see here I think "this is about getting an iterator of Arrow fields out of the schema, based on whether they're influx tags, fields or timestamp".

If that's the case, I wonder if a lot of this PR could be replaced with something like this on the `Schema` type.

```rust
    pub fn tags_iter(&self) -> impl Iterator<Item = &ArrowField> {
        (0..self.len()).filter_map(move |idx| {
            let (t, field) = self.field(idx);
            if matches!(t, Some(InfluxColumnType::Tag)) {
                return Some(field);
            }
            None
        })
    }

    pub fn fields_iter(&self) -> impl Iterator<Item = &ArrowField> {
        (0..self.len()).filter_map(move |idx| {
            let (t, field) = self.field(idx);
            if matches!(t, Some(InfluxColumnType::Field(_))) {
                return Some(field);
            }
            None
        })
    }
```

File: data_types/src/schema.rs [hunk 2]
```diff
@@ -567,30 +589,131 @@ impl From<&InfluxColumnType> for ArrowDataType {
...
+{
+    type Item = (Option<InfluxColumnType>, &'a ArrowField);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(idx) = self.next_idx() {
+            let (influx_colum_type, field) = self.schema.field(idx);
+            if self.filter.passes(influx_colum_type, field) {
+                return Some((influx_colum_type, field));
+            }
+        }
```
Comment (e-dard) [inline data_types/src/schema.rs:714]:
```suggestion
    fn next(&mut self) -> Option<Self::Item> {
        while let Some(idx) = self.next_idx() {
            let (influx_column_type, field) = self.schema.field(idx);
            if self.filter.passes(influx_column_type, field) {
                return Some((influx_column_type, field));
            }
        }
```

**Review by e-dard — COMMENTED** [review]
Code looks good to me but I have a broader question in the PR discussion.

File: data_types/src/schema.rs [hunk 3]
```diff
@@ -298,10 +298,32 @@ impl Schema {
...
+            inner: SchemaIter::new(self),
+        }
+    }
+
+    /// Returns an iterator of `&Field` for all the timestamp columns
+    /// of this schema, in order. At the time of writing there should
+    /// be only one or 0 such columns
+    pub fn time_iter(&self) -> ArrowFieldIter<'_, TimeColumns> {
+        ArrowFieldIter {
+            inner: SchemaIter::new(self),
```
Comment (NGA-TRAN) [inline data_types/src/schema.rs:326]: These are cool and exactly what I wished for when I implemented Optimize Selection Column

**Comment by alamb** [comment]
> If that's the case, I wonder if a lot of this PR could be replaced with something like this on the Schema type.

@e-dard   that is a great idea. Let me see if I can make that work

**Comment by alamb** [comment]
OMG @e-dard  -- the suggestion to use `impl ` to define the output type is *amazing* -- I didn't realize you could do that (I thought you could only do it on input parameters) 👍

**Comment by alamb** [comment]
After @e-dard 's wonderful suggestion to use `impl Trait`, implemented in https://github.com/influxdata/influxdb_iox/pull/879/commits/b852d7e7a91409409f390fd3d1098828db6e77c0 this PR has devolved into mostly tests

**Review by e-dard — APPROVED** [review]
LGTM. 👍

**Pull request merged**